src/auslib/util/versions.py

```diff
@@ -43,10 +43,49 @@ class AncientMozillaVersion(StrictVersion):
     )
 
 
+class GlobVersionTuple(tuple):
```

Comment:
I'm probably being dense here, but why does this need to exist? AFAICT it behaves the same a regular tuple:
```
>>> (1, 3, 5) == (1, 3, 5)
True
>>> 
>>> class GlobVersionTuple(tuple):
    def __eq__(self, value):
        for i, j in enumerate(self):
            if j != value[i]:
                return False
        return True

    def __ne__(self, value):
        return not self.__eq__(value)
>>> 
>>> 
>>> gvt = GlobVersionTuple((1,3,5))
>>> gvt2 = GlobVersionTuple((1,3,5))
>>> gvt == gvt2
True
```

Reply:
We're comparing a version against a glob tuple. So

`GlobVersion("89.8.*")` will get a `GlobVersionTuple` of `(89, 8)`. `(89, 8) != (89, 8, 1)` even though `89.8.*` should match `89.8.1`. So basically `GlobVersionTuple` makes sure `version_tuple[:len(glob_tuple)] == glob_tuple`, but that leads to infinite recursion in the `__eq__` function, so instead I enumerate the `glob_tuple` and make sure all of its version numbers match the initial version numbers in the `version_tuple`.

Reply:
Oh, I see. So the left and right side are not treated the same -- we ignore any items from the right that don't have an equivalent on the left. I find this a bit weird (AFAICT it means that foo == bar may not be equivalent to bar == foo), but I can live with this for now as long as it's documented. (If I'm wrong about any of this, please let me know.)

Reply:
I think it works both ways?

```
#!/usr/bin/env python

import re
from distutils.version import LooseVersion, StrictVersion

class GlobVersionTuple(tuple):
    def __eq__(self, value):
        # If we're 78.8.*, don't match 78.8. Also avoid a KeyError below
        if len(value) < len(self):
            return False
        for i, j in enumerate(self):
            if j != value[i]:
                return False
        return True

    def __ne__(self, value):
        return not self.__eq__(value)


class GlobVersion(StrictVersion):
    """A version class that supports Firefox versions 5.0 and up, which ends
    with a glob `*`. Not really a StrictVersion at all, but it needs to be
    to compare with other MozillaVersions."""

    version_re = re.compile(
        r"""^(\d+) \. (\d+\.\*|\*)$""",
        re.VERBOSE,
    )
    prerelease = None

    def parse(self, vstring):
        self.vstring = vstring
        match = self.version_re.match(vstring)
        if not match:
            raise ValueError("invalid version number '%s'" % vstring)
        parts = [int(i) for i in vstring.split(".") if i != "*"]
        self.version = GlobVersionTuple(parts)

    def __eq__(self, value):
        return str(value).startswith(self.vstring.rstrip("*"))

g = GlobVersionTuple([7, 9])

print(g == (7, 9, 2))
print(g == (7, 9))
print((7,9,2) == g)
print((7,9) == g)
```

prints `True` four times.

Reply:
Just to be sure, I tested

```
print(g == StrictVersion("7.9.2").version)
print(g == StrictVersion("7.9.0").version)
print(StrictVersion("7.9.2").version == g)
print(StrictVersion("7.9.0").version == g)
```

and those all printed `True` as well, and

```
print(StrictVersion("7.90.2").version == g)
print(g == StrictVersion("7.90.2").version)
```

returns False. Want me to add some parametrized tests here?

Reply:
Added more tests in https://github.com/mozilla-releng/balrog/pull/1924/commits/f13bba07e065044ac8aad7194ee93899ffe85f52 .

Reply:
ok, clearly i'm misreading the code! ty for adding those tests